core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -0,0 +1,231 @@
+import org.infinispan.xsite.irac.DefaultIracManager;
+import org.infinispan.xsite.irac.IracExecutor;
+
+/**
+ * A default implementation for {@link IracTombstoneManager}.
+ * <p>
+ * This class is responsible to keep track of the tombstones for the IRAC algorithm. Tombstones are used when a key is
+ * removed but its metadata is necessary to detect possible conflicts in this and remote sites. When all sites have
+ * updated the key, the tombstone can be removed.
+ * <p>
+ * Tombstones are removed periodically in the background.
+ *
+ * @since 14.0
+ */
+@Scope(Scopes.NAMED_CACHE)
+public class DefaultIracTombstoneManager implements IracTombstoneManager {
+
+   @Inject
+   DistributionManager distributionManager;
+   @Inject
+   RpcManager rpcManager;
+   @Inject
+   CommandsFactory commandsFactory;
+   private final Map<Object, TombstoneData> tombstoneMap;
+   private final IracExecutor iracExecutor;
+   private final Collection<XSiteBackup> asyncBackups;
+
+   public DefaultIracTombstoneManager(Configuration configuration) {
+      this.iracExecutor = new IracExecutor(this::performCleanup);
+      this.asyncBackups = DefaultIracManager.asyncBackups(configuration);
+      this.tombstoneMap = new ConcurrentHashMap<>();
+   }
+
+   @Inject
+   public void inject(@ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService executorService,
+                      @ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor blockingExecutor) {
+      // using the inject method here in order to decrease the class size
+      iracExecutor.setBackOff(ExponentialBackOff.NO_OP);
+      iracExecutor.setExecutor(blockingExecutor);
+      // TODO configure? https://issues.redhat.com/browse/ISPN-13446
+      executorService.scheduleAtFixedRate(this::startCleanupTombstone, 30, 30, TimeUnit.SECONDS);
+   }
+
+   public void storeTombstone(int segment, Object key, IracMetadata metadata) {
+      tombstoneMap.put(key, new TombstoneData(segment, metadata));
+   }
+
+   @Override
+   public void storeTombstoneIfAbsent(int segment, Object key, IracMetadata metadata) {
+      if (metadata == null) {
+         return;
+      }
+      tombstoneMap.putIfAbsent(key, new TombstoneData(segment, metadata));
+   }
+
+   @Override
+   public IracMetadata getTombstone(Object key) {
+      TombstoneData data = tombstoneMap.get(key);
+      return data == null ? null : data.getMetadata();
+   }
+
+   @Override
+   public void removeTombstone(Object key, IracMetadata iracMetadata) {
+      if (iracMetadata == null) {
+         return;
+      }
+      remove(key, new TombstoneData(-1, iracMetadata));
+   }
+
+   @Override
+   public void removeTombstone(Object key) {
+      tombstoneMap.remove(key);
+   }
+
+
+   @Override
+   public boolean isEmpty() {
+      return tombstoneMap.isEmpty();
+   }
+
+   @Override
+   public int size() {
+      return tombstoneMap.size();
+   }
+
+   public void startCleanupTombstone() {
```

Comment:
Can be inlined

Reply:
what do you mean?

Reply:
I meant the tests can use `IracExecutor.run()` directly and the method could be removed. Originally I was also thinking of the call in `start()`, but I see you've already replaced it there.